pkg/bootstrap/platform/gcp.go

```diff
@@ -207,19 +210,52 @@ func parseGCPMetadata() (pid, npid, cluster, location string) {
 	return envPid, envNpid, envCluster, envLocation
 }
 
+func isZone(z string) bool {
+	return zonePattern.MatchString(z)
+}
+
 // Converts a GCP zone into a region.
 func zoneToRegion(z string) (string, error) {
-	// Zones are in the form <region>-<zone_suffix>, so capture everything but the suffix.
-	re := regexp.MustCompile("(.*)-.*")
-	m := re.FindStringSubmatch(z)
-	if len(m) != 2 {
-		return "", fmt.Errorf("unable to extract region from GCP zone: %s", z)
+	if !isZone(z) {
+		return "", fmt.Errorf("invalid zone: %q", z)
 	}
-	return m[1], nil
+	zoneSuffixIdx := strings.LastIndex(z, "-")
+	return z[:zoneSuffixIdx], nil
 }
 
 // Locality returns the GCP-specific region and zone.
 func (e *gcpEnv) Locality() *core.Locality {
+	if e.locality != nil {
+		return e.locality
+	}
+	l := locality()
+	e.Lock()
+	e.locality = l
+	e.Unlock()
+	return e.locality
+}
+
+func locality() *core.Locality
```

Comment:
is there any reason we don't use Metadata() here? Is zone not returned in the same info?

Reply:
I didn't want to call `Metadata()` because it looks up potentially different location info ("cluster-location" could be region, I think) and it makes additional calls to the MD service -- so, if the calls to the MD service are what we are trying to avoid, that seems somewhat problematic.

Thoughts?